FLT/HIMExperiments/module_topology.lean

```diff
@@ -213,6 +216,111 @@ noncomputable def LinearMap.add (M : Type*) [AddCommGroup M] [Module A M] :
   map_add' _ _ := add_add_add_comm _ _ _ _
   map_smul' _ _ := (DistribSMul.smul_add _ _ _).symm
 
+/-- Basis.repr in the first variable as a linear map. -/
+noncomputable def LinearMap.basis₁ (M : Type*) (ι : Type*) [Finite ι] [AddCommGroup M] [Module A M]
+  (b : Basis ι A M) (i : ι): M →ₗ[A] A where
+  toFun m := b.repr m i
+  map_add' := by simp only [map_add, Finsupp.coe_add, Pi.add_apply, implies_true]
+  map_smul' _ _  := by simp only [LinearMapClass.map_smul, Finsupp.coe_smul, Pi.smul_apply,
+    smul_eq_mul, RingHom.id_apply]
+
+noncomputable def LinearMap.prodfst (M N : Type*) [AddCommGroup M] [Module A M] [AddCommGroup N]
+  [Module A N] : M × N →ₗ[A] M where
+  toFun := Prod.fst
+  map_add' _ _ := rfl
+  map_smul' _ _ := rfl
+
+noncomputable def LinearMap.prodsnd (M N : Type*) [AddCommGroup M] [Module A M] [AddCommGroup N]
+  [Module A N] : M × N →ₗ[A] N where
+  toFun := Prod.snd
+  map_add' _ _ := rfl
+  map_smul' _ _:= rfl
+
+instance Module.instCommAdd {P : Type*} [AddCommGroup P] [Module A P]:
+@ContinuousAdd P (Module.topology A) _ := by
+  apply @ContinuousAdd.mk _ (topology A)
+  rw [Module.prod_canonical A]
+  exact Module.continuous_linear A (LinearMap.add A P)
+
+variable [Module.Finite A M] [Module.Free A M] [Module.Finite A N] [Module.Free A N]
+
+
+-- I really shouldn't start with iA instead of topology A here.
+-- But I need it like this in the next proof and don't know how to rewrite it there
+-- Sorry for leaving it like this...
+instance Module.instContinuousSMul : @ContinuousSMul A M _ _ (topology A) := by
+  let _τM : TopologicalSpace M := Module.topology A
+  apply @ContinuousSMul.mk A M _ _ (topology A)
+  let ι := Free.ChooseBasisIndex A M
+  have b : Basis ι A M := Free.chooseBasis A M
+  suffices Continuous fun (p : A × M) ↦ ∑ i : ι, p.1 • b.repr p.2 i • b i by
+    convert this using 1
+    simp_rw [← Finset.smul_sum, Basis.sum_repr]
```

Comment:
```suggestion
    simpa [← Finset.smul_sum, Basis.sum_repr]
```
Avoid `convert` if you can, it's quite fragile. `simpa` is great for this. It simplifies the goal and `this` and then tries to prove the goal with `exact this`.